tests/unittests/tests-phydat/tests-phydat.c

```diff
@@ -16,44 +17,77 @@
-    /* Check that the result was rescaled to 10044e-5 */
-    /* The scaled number is rounded toward zero */
-    TEST_ASSERT_EQUAL_INT(1, res);
-    TEST_ASSERT_EQUAL_INT(UNIT_V, dat.unit);
-    TEST_ASSERT_EQUAL_INT(-5, dat.scale);
-    TEST_ASSERT_EQUAL_INT( 10044, dat.val[0]);
-    /* Fit the next value in the phydat vector */
-    res = phydat_fit(&dat, val1, 1, res);
-    TEST_ASSERT_EQUAL_INT(2, res);
-    TEST_ASSERT_EQUAL_INT(UNIT_V, dat.unit);
-    TEST_ASSERT_EQUAL_INT(-4, dat.scale);
-    TEST_ASSERT_EQUAL_INT(  1004, dat.val[0]);
-    TEST_ASSERT_EQUAL_INT( 20009, dat.val[1]);
-    /* Fit the third value in the phydat vector */
-    res = phydat_fit(&dat, val2, 2, res);
-    TEST_ASSERT_EQUAL_INT(3, res);
-    TEST_ASSERT_EQUAL_INT(UNIT_V, dat.unit);
-    TEST_ASSERT_EQUAL_INT(-3, dat.scale);
-    TEST_ASSERT_EQUAL_INT(   100, dat.val[0]);
-    TEST_ASSERT_EQUAL_INT(  2000, dat.val[1]);
-    TEST_ASSERT_EQUAL_INT( 30000, dat.val[2]);
-    /* Overwrite the second value in the phydat vector */
-    res = phydat_fit(&dat, val4, 1, res);
-    TEST_ASSERT_EQUAL_INT(3, res);
-    TEST_ASSERT_EQUAL_INT(UNIT_V, dat.unit);
-    TEST_ASSERT_EQUAL_INT(-3, dat.scale);
-    TEST_ASSERT_EQUAL_INT(   100, dat.val[0]);
-    TEST_ASSERT_EQUAL_INT(  1234, dat.val[1]);
-    TEST_ASSERT_EQUAL_INT( 30000, dat.val[2]);
+    /* Input values for each test: */
+    static const int32_t values[][3] = {
+        { 100445, -1, -1 },
+        { -5, 2000954, 3 },
+        { 30000449, -30000450, 30000500 },
+        { -30000449, -30000499, -30000500 },
+        { 0, 0, 1234567 },
+        { 32767, 32767, 32767 },
+        { 32766, 32766, 32766 },
+        { -32767, -32767, -32767 },
+        { -32766, -32766, -32766 },
+    };
+    static const int8_t scales[] = {
+        -6,
+        42,
+        0,
+        -1,
+        5,
+        0,
+        0,
+        0,
+        0
+    };
+    static const unsigned int dims[] = {
+        1,
+        2,
+        3,
+        3,
+        3,
+        3,
+        3,
+        3,
+        3,
+    };
+    static const uint8_t units[] = {
+        UNIT_V,
+        UNIT_A,
+        UNIT_NONE,
+        UNIT_LUX,
+        UNIT_M,
+        UNIT_NONE,
+        UNIT_NONE,
+        UNIT_NONE,
+        UNIT_NONE,
+    };
+    /* Expected output values for each test: */
+    static const phydat_t expected[] = {
+        { .val = {  10045,     -1,     -1 }, .unit = UNIT_V,    .scale = -5 },
+        { .val = {      0,  20010,     -1 }, .unit = UNIT_A,    .scale = 44 },
+        { .val = {  30000, -30000,  30001 }, .unit = UNIT_NONE, .scale =  3 },
+        { .val = { -30000, -30000, -30001 }, .unit = UNIT_LUX,  .scale =  2 },
+        { .val = {      0,      0,  12346 }, .unit = UNIT_M,    .scale =  7 },
+        { .val = {   3277,   3277,   3277 }, .unit = UNIT_NONE, .scale =  1 },
+        { .val = {  32766,  32766,  32766 }, .unit = UNIT_NONE, .scale =  0 },
+        { .val = {  -3277,  -3277,  -3277 }, .unit = UNIT_NONE, .scale =  1 },
```

Comment:
Wrong behaviour IMO.
-32767 fits inside phydat_t::dat

Also, please add tests for -32768 (expect -32768), -32769 (expect -3277, scale +1), 32767 (expect 32767), 32768 (expect 3277, scale 1)

Reply:
This was actually a deliberate implementation decision for two reasons:
- Both positive and negative numbers can be treated identical if treating `-32768` as out of range
- Special handling of the scale estimation can be skipped if treating `32767` (and thus also `-32767`) as out of range.

The special case for scale estimation is when the number is 32767 and has to be rounded up. By treating only `[-32766, 32766]` as target range no special treatment is required, which makes the code shorter, the ROM smaller and faster. Also, in 99.99542% of random numbers the result is the some, in the remaining 0.00458% of the cases some precision is lost.

But let me think about it. Maybe there is an elegant solution to use the full range.

Reply:
Good arguments. It makes sense to introduce this limitation to save on code size and simplify the implementation if a rationale is clearly described in the module documentation.